### PR TITLE
feat(admin-ui): Filter reservations

### PR DESCRIPTION
**Pull request by pKorsholm**
**What**
- Introduce reservation filters in admin

Fixes CORE-1387

### CODE REVIEW

File: packages/medusa/src/api/routes/admin/reservations/create-reservation.ts [hunk 1]
```diff
@@ -121,6 +121,7 @@ export default async (req, res) => {
  */
 export class AdminPostReservationsReq {
   @IsString()
+  @IsOptional()
```
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/reservations/create-reservation.ts]: question: Does that mean that we support reservations that are not related to an item?

File: packages/medusa/src/api/routes/admin/reservations/index.ts [hunk 2]
```diff
@@ -67,6 +69,25 @@ export type AdminReservationsRes = {
...
+ *   - type: object
+ *     properties:
+ *       line_item:
+ *         description: optional line item
+ *         $ref: "#/components/schemas/LineItem"
+ *       inventory_item:
+ *         description: inventory item from inventory module
+ *         $ref: "#/components/schemas/InventoryItemDTO"
+ */
+export type ExtendedReservationItem = ReservationItemDTO & {
```
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/reservations/index.ts]: suggestion: Should we move that to the types package?

File: packages/medusa/src/api/routes/admin/reservations/list-reservations.ts [hunk 3]
```diff
@@ -110,12 +116,60 @@ import {
...
+
+  if (includeInventoryItems) {
+    relationsPredicates.push((r) => r !== "inventory_item")
+  }
+
+  if (relationsPredicates.length) {
+    listConfig.relations = listConfig.relations?.filter((r) =>
+      relationsPredicates.every((p) => p(r))
+    )
+  }
```
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/reservations/list-reservations.ts]: todo: From my understanding, we just want to remove line_item and inventory_item from the relations, in that case I would suggest something simpler which avoid looping through the predicates and looping through the relations as well as making it more readable and easy to update if needed

```ts
const relations = new Set(listConfig.relations ?? [])

const includeItems = relations.delete("line_item")
const includeInventoryItems = relations.delete("inventory_item")

if (listConfig.relations?.length) {
  listConfig.relations = [...relations]
}

// ...
```

Wdyt?

File: packages/medusa/src/api/routes/admin/reservations/utils/join-inventory-items.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      },
+      {},
+      {
+        transactionManager: dependencies.manager,
+      }
+    )
+
+  const inventoryItemMap = new Map(inventoryItems.map((i) => [i.id, i]))
+
+  return await Promise.all(
```
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/reservations/utils/join-inventory-items.ts]: todo: rm Promise.all ans just a non async map should suffice

File: packages/medusa/src/api/routes/admin/reservations/utils/join-line-items.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    },
+    {
+      relations: ["order"],
+    }
+  )
+
+  const lineItemMap = new Map(lineItems.map((i) => [i.id, i]))
+
+  return await Promise.all(
+    reservations.map(async (reservation) => {
```
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/reservations/utils/join-line-items.ts]: todo: same as the previous comment

**Review by adrien2p — CHANGES_REQUESTED** [review]
I ve only reviewed the backenmd for now

File: packages/medusa/src/api/routes/admin/reservations/list-reservations.ts (hunk 3 above)
Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/reservations/list-reservations.ts]: One stone two birds 😅 let me know what you think

File: packages/medusa/src/api/routes/admin/reservations/create-reservation.ts (hunk 1 above)
Comment (pKorsholm) [reply packages/medusa/src/api/routes/admin/reservations/create-reservation.ts]: Yes, "Custom reservations" that can exist outside of orders

File: packages/medusa/src/api/routes/admin/reservations/list-reservations.ts (hunk 3 above)
Comment (pKorsholm) [reply packages/medusa/src/api/routes/admin/reservations/list-reservations.ts]: sure

**Comment by pKorsholm** [comment]
@adrien2p I addressed your feedback in #4081 as that's got to do with the backend and this should only be the admin ui. I think updating the other branch pushed a bunch of unrelated changes to the files, but your comments are fixed in the other pr.

**Comment by adrien2p** [comment]
> @adrien2p I addressed your feedback in #4081 as that's got to do with the backend and this should only be the admin ui. I think updating the other branch pushed a bunch of unrelated changes to the files, but your comments are fixed in the other pr.

sounds good 🚀

File: packages/admin-ui/ui/src/components/atoms/date-picker/date-picker.tsx [hunk 6]
```diff
@@ -11,10 +11,11 @@ import InputHeader from "../../fundamentals/input-header"
 import CustomHeader from "./custom-header"
 import { DateTimePickerProps } from "./types"
 
-const getDateClassname = (d: Date, tempDate: Date) => {
+const getDateClassname = (d: Date, tempDate: Date, greyPast: boolean) => {
```
Comment (olivermrbl) [inline packages/admin-ui/ui/src/components/atoms/date-picker/date-picker.tsx]: todo: More intuitive naming for `greyPast`. Think you will need the context to understand this one.

Comment (olivermrbl) [reply packages/admin-ui/ui/src/components/atoms/date-picker/date-picker.tsx]: Should this have a default value assigned here instead of in the calendar component to stay backward compatible?

File: packages/admin-ui/ui/src/components/templates/reservations-table/index.tsx [hunk 7]
```diff
@@ -222,16 +289,19 @@ const ReservationsTable: React.FC<ReservationsTableProps> = () => {
         isLoading={isLoading}
       >
         <Table
-          enableSearch
+          // enableSearch
```
Comment (olivermrbl) [inline packages/admin-ui/ui/src/components/templates/reservations-table/index.tsx]: todo:
```suggestion
```

File: packages/admin-ui/ui/src/components/molecules/select/next-select/components/single-value.tsx [hunk 8]
```diff
@@ -13,28 +16,50 @@ const SingleValue = <
...
+    if (!e || !(e.offsetParent as HTMLDivElement)?.offsetWidth) {
+      return false
+    }
+
+    return (e.offsetParent as HTMLDivElement).offsetWidth < e.scrollWidth
+  }
+
+  const toolTip = !getValue().length
+    ? null
+    : ((getValue()?.[0] as { label: string })?.label as string) ?? null
```
Comment (olivermrbl) [inline packages/admin-ui/ui/src/components/molecules/select/next-select/components/single-value.tsx]: comment: This looks a bit funky. If possible, we should make this a tiny bit more readable 😄

**Review by olivermrbl — CHANGES_REQUESTED** [review]
epic work @pKorsholm !! 🚀 

only minuscule todos

**Review by olivermrbl — CHANGES_REQUESTED** [review]
epic work @pKorsholm !! 🚀 

only minuscule todos

**Comment by olivermrbl** [comment]
Also, missing a changeset :)

File: packages/admin-ui/ui/src/components/atoms/date-picker/date-picker.tsx (hunk 6 above)
Comment (pKorsholm) [reply packages/admin-ui/ui/src/components/atoms/date-picker/date-picker.tsx]: went for `greyPastDates`, wdyt? the notion is a bit hard to communicate but I see the confusion that could arise with just `greyPast`.

Assigned default values here and in the component 😄

File: packages/admin-ui/ui/src/components/molecules/select/next-select/components/single-value.tsx (hunk 8 above)
Comment (pKorsholm) [reply packages/admin-ui/ui/src/components/molecules/select/next-select/components/single-value.tsx]: created a short util method instead to extract this in the component for readability

**Comment by olivermrbl** [comment]
@adrien2p will you check if your requested changes have been addressed?
